lms/djangoapps/completion/models.py

```diff
@@ -0,0 +1,120 @@
+        Update the completed value for the specified record.
+
+        Parameters:
+            * user (django.contrib.auth.models.User): The user for whom the
+              completion is being submitted.
+            * course_key (opaque_keys.edx.keys.CourseKey): The course in
+              which the submitted block is found.
+            * block_key (opaque_keys.edx.keys.UsageKey): The block that has had
+              its completion changed.
+            * completed (float in range [0.0, 1.0]): The fractional completion
+              value of the block (0.0 = incomplete, 1.0 = complete).
+
+        Return Value:
+            (BlockCompletion, bool): A tuple comprising the created or updated
+            BlockCompletion object and a boolean value indicating whether the value
+
+        Raises:
+
+            ValueError:
+                If the wrong type is passed for one of the parameters.
+
+            django.core.exceptions.ValidationError:
+                If a float is passed that is not between 0.0 and 1.0.
+
+            django.db.DatabaseError:
+                If there was a problem getting, creating, or updating the
+                BlockCompletion record in the database.
+
+                This will also be a more specific error, as described here:
+                https://docs.djangoproject.com/en/1.11/ref/exceptions/#database-exceptions.
+                IntegrityError and OperationalError are relatively common
+                subclasses.
+        """
+
+        # Raise ValueError to match normal django semantics for wrong type of field.
+        if not isinstance(course_key, CourseKey):
+            raise ValueError(
+                "course_key must be an instance of `opaque_keys.edx.keys.CourseKey`.  Got {}".format(type(course_key))
+            )
+        try:
+            block_type = block_key.block_type
+        except AttributeError:
+            raise ValueError(
+                "block_key must be an instance of `opaque_keys.edx.keys.UsageKey`.  Got {}".format(type(block_key))
+            )
+
+        obj, isnew = self.get_or_create(
+            user=user,
+            course_key=course_key,
+            block_type=block_type,
+            block_key=block_key,
+            defaults={'completed': completed},
+        )
+        if not isnew and obj.completed != completed:
+            obj.completed = completed
+            obj.full_clean()
+            obj.save()
+        return obj, isnew
+
+
+class BlockCompletion(TimeStampedModel, models.Model):
+    """
+    Track completion of completable blocks.
+    """
+    user = models.ForeignKey(User)
+    course_key = CourseKeyField(max_length=255)
+    block_key = UsageKeyField(max_length=255)
+    block_type = models.CharField(max_length=64)
+    completed = models.FloatField(validators=[validate_ratio])
+
+    objects = BlockCompletionManager()
+
+    class Meta(object):
+        index_together = [
+            ('course_key', 'block_type', 'user'),
+        ]
```

Comment:
This will work as an index to retrieve any of the following:
* all completions in a course
* all completions of a particular type in a course
* all completions of a particular type for a given user in a course